-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34943/24764

ERROR: Build errors found during clang-tidy run.
```
RecoMET/METFilters/interface/EcalBoundaryInfoCalculator.h:270:47: error: functions that differ only in their return type cannot be overloaded [clang-diagnostic-error]
    std::unique_ptr<CaloNavigator<EcalDetId>> theEcalNav();
                                              ^
RecoMET/METFilters/interface/EcalBoundaryInfoCalculator.h:548:21: note: in instantiation of member function 'EcalBoundaryInfoCalculator<EEDetId>::initializeEcalNavigator' requested here
--
RecoMET/METFilters/interface/EcalBoundaryInfoCalculator.h:272:18: error: non-object type 'std::unique_ptr<CaloNavigator<EBDetId>> ()' is not assignable [clang-diagnostic-error]
      theEcalNav = std::make_unique<CaloNavigator<EcalDetId>>(
                 ^
RecoMET/METFilters/interface/EcalBoundaryInfoCalculator.h:548:21: note: in instantiation of member function 'EcalBoundaryInfoCalculator<EBDetId>::initializeEcalNavigator' requested here
--
RecoMET/METFilters/interface/EcalBoundaryInfoCalculator.h:281:12: error: no viable conversion from returned value of type 'std::unique_ptr<CaloNavigator<EBDetId>> ()' to function return type 'std::unique_ptr<CaloNavigator<EBDetId>>' [clang-diagnostic-error]
    return theEcalNav;
           ^
/cvmfs/cms-ib.cern.ch/nweek-02694/slc7_amd64_gcc900/external/gcc/9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../include/c++/9.3.0/bits/unique_ptr.h:250:12: note: candidate constructor template not viable: no known conversion from 'std::unique_ptr<CaloNavigator<EBDetId>> ()' to 'std::nullptr_t' (aka 'nullptr_t') for 1st argument
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/RecoMET/METAlgorithms/interface/CSCHaloAlgo.h
+++ b/RecoMET/METAlgorithms/interface/CSCHaloAlgo.h
@@ -89,7 +89,7 @@ namespace edm {
 
 class CSCHaloAlgo {
 public:
-  CSCHaloAlgo();
+  CSCHaloAlgo(edm::ConsumesCollector iC);
   ~CSCHaloAlgo() {}
   reco::CSCHaloData Calculate(const CSCGeometry& TheCSCGeometry,
                               edm::Handle<reco::MuonCollection>& TheCosmicMuons,
@@ -170,6 +170,7 @@ private:
   float et_thresh_rh_ee, dphi_thresh_segvsrh_ee, dr_lowthresh_segvsrh_ee, dr_highthresh_segvsrh_ee,
       dt_lowthresh_segvsrh_ee, dt_highthresh_segvsrh_ee;
 
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> geoToken_;
   const CaloGeometry* geo_;
   const HcalGeometry* hgeo_;
   math::XYZPoint getPosition(const DetId& id, reco::Vertex::Point vtx);

--- a/RecoMET/METAlgorithms/src/CSCHaloAlgo.cc
+++ b/RecoMET/METAlgorithms/src/CSCHaloAlgo.cc
@@ -14,7 +14,7 @@ using namespace edm;
 namespace {
   constexpr float c_cm_per_ns = 29.9792458;
 };
-CSCHaloAlgo::CSCHaloAlgo() : geo_(nullptr), hgeo_(nullptr) {
+CSCHaloAlgo::CSCHaloAlgo(edm::ConsumesCollector iC) : geoToken_(iC.esConsumes()), geo_(nullptr), hgeo_(nullptr) {
   deta_threshold = 0.;
   min_inner_radius = 0.;
   max_inner_radius = 9999.;
@@ -84,9 +84,7 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
   bool ECALEmatched = false;
   bool HCALmatched = false;
 
-  edm::ESHandle<CaloGeometry> pGeo;
-  TheSetup.get<CaloGeometryRecord>().get(pGeo);
-  geo_ = pGeo.product();
+  geo_ = &TheSetup.getData(geoToken_);
   hgeo_ = dynamic_cast<const HcalGeometry*>(geo_->getSubdetectorGeometry(DetId::Hcal, 1));
 
   //}

--- a/RecoMET/METPUSubtraction/interface/PFMETAlgorithmMVA.h
+++ b/RecoMET/METPUSubtraction/interface/PFMETAlgorithmMVA.h
@@ -11,9 +11,11 @@
  */
 
 #include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 
 #include "CondFormats/GBRForest/interface/GBRForest.h"
+#include "CondFormats/DataRecord/interface/GBRWrapperRcd.h"
 
 #include "DataFormats/Candidate/interface/Candidate.h"
 #include "DataFormats/METReco/interface/MET.h"
@@ -29,7 +31,7 @@
 
 class PFMETAlgorithmMVA {
 public:
-  PFMETAlgorithmMVA(const edm::ParameterSet& cfg);
+  PFMETAlgorithmMVA(const edm::ParameterSet& cfg, edm::ConsumesCollector iC);
   ~PFMETAlgorithmMVA();
 
   void initialize(const edm::EventSetup&);
@@ -56,7 +58,6 @@ public:
 private:
   const std::string updateVariableNames(std::string input);
   const GBRForest* loadMVAfromFile(const edm::FileInPath& inputFileName, const std::string& mvaName);
-  const GBRForest* loadMVAfromDB(const edm::EventSetup& es, const std::string& mvaName);
 
   const float evaluateU();
   const float evaluateDPhi();
@@ -69,6 +70,11 @@ private:
   std::string mvaNameDPhi_;
   std::string mvaNameCovU1_;
   std::string mvaNameCovU2_;
+
+  edm::ESGetToken<GBRForest, GBRWrapperRcd> mvaTokenU_;
+  edm::ESGetToken<GBRForest, GBRWrapperRcd> mvaTokenDPhi_;
+  edm::ESGetToken<GBRForest, GBRWrapperRcd> mvaTokenCovU1_;
+  edm::ESGetToken<GBRForest, GBRWrapperRcd> mvaTokenCovU2_;
 
   int mvaType_;
   bool hasPhotons_;

--- a/RecoMET/METPUSubtraction/plugins/PFMETProducerMVA.cc
+++ b/RecoMET/METPUSubtraction/plugins/PFMETProducerMVA.cc
@@ -7,7 +7,8 @@ const double dR2Min = 0.01 * 0.01;
 const double dR2Max = 0.5 * 0.5;
 const double dPtMatch = 0.1;
 
-PFMETProducerMVA::PFMETProducerMVA(const edm::ParameterSet& cfg) : mvaMEtAlgo_(cfg), mvaMEtAlgo_isInitialized_(false) {
+PFMETProducerMVA::PFMETProducerMVA(const edm::ParameterSet& cfg)
+    : mvaMEtAlgo_(cfg, consumesCollector()), mvaMEtAlgo_isInitialized_(false) {
   srcCorrJets_ = consumes<reco::PFJetCollection>(cfg.getParameter<edm::InputTag>("srcCorrJets"));
   srcUncorrJets_ = consumes<reco::PFJetCollection>(cfg.getParameter<edm::InputTag>("srcUncorrJets"));
   srcJetIds_ = consumes<edm::ValueMap<float> >(cfg.getParameter<edm::InputTag>("srcMVAPileupJetId"));

--- a/RecoMET/METPUSubtraction/src/PFMETAlgorithmMVA.cc
+++ b/RecoMET/METPUSubtraction/src/PFMETAlgorithmMVA.cc
@@ -3,10 +3,6 @@
 #include "FWCore/ParameterSet/interface/FileInPath.h"
 #include "FWCore/Utilities/interface/Exception.h"
 
-#include "FWCore/ServiceRegistry/interface/Service.h"
-#include "FWCore/Framework/interface/ESHandle.h"
-#include "CondFormats/DataRecord/interface/GBRWrapperRcd.h"
-
 #include "DataFormats/METReco/interface/CommonMETData.h"
 
 #include <TFile.h>
@@ -112,13 +108,7 @@ const GBRForest* PFMETAlgorithmMVA::loadMVAfromFile(const edm::FileInPath& input
   return mva;
 }
 
-const GBRForest* PFMETAlgorithmMVA::loadMVAfromDB(const edm::EventSetup& es, const std::string& mvaName) {
-  edm::ESHandle<GBRForest> mva;
-  es.get<GBRWrapperRcd>().get(mvaName, mva);
-  return mva.product();
-}
-
-PFMETAlgorithmMVA::PFMETAlgorithmMVA(const edm::ParameterSet& cfg)
+PFMETAlgorithmMVA::PFMETAlgorithmMVA(const edm::ParameterSet& cfg, edm::ConsumesCollector iC)
     : utils_(cfg),
       mvaReaderU_(nullptr),
       mvaReaderDPhi_(nullptr),
@@ -127,7 +117,19 @@ PFMETAlgorithmMVA::PFMETAlgorithmMVA(const edm::ParameterSet& cfg)
       cfg_(cfg) {
   mvaType_ = kBaseline;
 
+  edm::ParameterSet cfgInputRecords = cfg_.getParameter<edm::ParameterSet>("inputRecords");
+  mvaNameU_ = cfgInputRecords.getParameter<std::string>("U");
+  mvaNameDPhi_ = cfgInputRecords.getParameter<std::string>("DPhi");
+  mvaNameCovU1_ = cfgInputRecords.getParameter<std::string>("CovU1");
+  mvaNameCovU2_ = cfgInputRecords.getParameter<std::string>("CovU2");
+
   loadMVAfromDB_ = cfg.getParameter<bool>("loadMVAfromDB");
+  if (loadMVAfromDB_) {
+    mvaTokenU_ = iC.esConsumes(edm::ESInputTag("", mvaNameU_));
+    mvaTokenDPhi_ = iC.esConsumes(edm::ESInputTag("", mvaNameDPhi_));
+    mvaTokenCovU1_ = iC.esConsumes(edm::ESInputTag("", mvaNameCovU1_));
+    mvaTokenCovU2_ = iC.esConsumes(edm::ESInputTag("", mvaNameCovU2_));
+  }
 }
 
 PFMETAlgorithmMVA::~PFMETAlgorithmMVA() {
@@ -141,17 +143,11 @@ PFMETAlgorithmMVA::~PFMETAlgorithmMVA() {
 
 //-------------------------------------------------------------------------------
 void PFMETAlgorithmMVA::initialize(const edm::EventSetup& es) {
-  edm::ParameterSet cfgInputRecords = cfg_.getParameter<edm::ParameterSet>("inputRecords");
-  mvaNameU_ = cfgInputRecords.getParameter<std::string>("U");
-  mvaNameDPhi_ = cfgInputRecords.getParameter<std::string>("DPhi");
-  mvaNameCovU1_ = cfgInputRecords.getParameter<std::string>("CovU1");
-  mvaNameCovU2_ = cfgInputRecords.getParameter<std::string>("CovU2");
-
   if (loadMVAfromDB_) {
-    mvaReaderU_ = loadMVAfromDB(es, mvaNameU_);
-    mvaReaderDPhi_ = loadMVAfromDB(es, mvaNameDPhi_);
-    mvaReaderCovU1_ = loadMVAfromDB(es, mvaNameCovU1_);
-    mvaReaderCovU2_ = loadMVAfromDB(es, mvaNameCovU2_);
+    mvaReaderU_ = &es.getData(mvaTokenU_);
+    mvaReaderDPhi_ = &es.getData(mvaTokenDPhi_);
+    mvaReaderCovU1_ = &es.getData(mvaTokenCovU1_);
+    mvaReaderCovU2_ = &es.getData(mvaTokenCovU2_);
   } else {
     edm::ParameterSet cfgInputFileNames = cfg_.getParameter<edm::ParameterSet>("inputFileNames");
 

--- a/RecoMET/METProducers/interface/PFMETProducer.h
+++ b/RecoMET/METProducers/interface/PFMETProducer.h
@@ -41,9 +41,9 @@
 #include "RecoMET/METAlgorithms/interface/SignPFSpecificAlgo.h"
 #include "RecoMET/METAlgorithms/interface/METSignificance.h"
 
+#include "CondFormats/DataRecord/interface/JetResolutionRcd.h"
+#include "CondFormats/DataRecord/interface/JetResolutionScaleFactorRcd.h"
 #include "JetMETCorrections/Modules/interface/JetResolution.h"
-
-#include "TVector.h"
 
 #include <string>
 
@@ -76,9 +76,10 @@ namespace cms {
 
     edm::EDGetTokenT<edm::View<reco::Jet>> jetToken_;
     std::vector<edm::EDGetTokenT<edm::View<reco::Candidate>>> lepTokens_;
-    std::string jetSFType_;
-    std::string jetResPtType_;
-    std::string jetResPhiType_;
+
+    edm::ESGetToken<JME::JetResolutionObject, JetResolutionScaleFactorRcd> jetSFToken_;
+    edm::ESGetToken<JME::JetResolutionObject, JetResolutionRcd> jetResPtToken_;
+    edm::ESGetToken<JME::JetResolutionObject, JetResolutionRcd> jetResPhiToken_;
     edm::EDGetTokenT<double> rhoToken_;
     bool applyWeight_;
     edm::EDGetTokenT<edm::ValueMap<float>> weightsToken_;

--- a/RecoMET/METProducers/src/CSCHaloDataProducer.cc
+++ b/RecoMET/METProducers/src/CSCHaloDataProducer.cc
@@ -13,7 +13,7 @@ using namespace edm;
 using namespace std;
 using namespace reco;
 
-CSCHaloDataProducer::CSCHaloDataProducer(const edm::ParameterSet& iConfig) {
+CSCHaloDataProducer::CSCHaloDataProducer(const edm::ParameterSet& iConfig) : CSCAlgo(consumesCollector()) {
   //Digi Level
   IT_L1MuGMTReadout = iConfig.getParameter<edm::InputTag>("L1MuGMTReadoutLabel");
 

--- a/RecoMET/METProducers/src/PFMETProducer.cc
+++ b/RecoMET/METProducers/src/PFMETProducer.cc
@@ -39,9 +39,9 @@ namespace cms {
         lepTokens_.push_back(mayConsume<edm::View<reco::Candidate>>(*it));
       }
 
-      jetSFType_ = iConfig.getParameter<std::string>("srcJetSF");
-      jetResPtType_ = iConfig.getParameter<std::string>("srcJetResPt");
-      jetResPhiType_ = iConfig.getParameter<std::string>("srcJetResPhi");
+      jetSFToken_ = esConsumes(edm::ESInputTag("", iConfig.getParameter<std::string>("srcJetSF")));
+      jetResPtToken_ = esConsumes(edm::ESInputTag("", iConfig.getParameter<std::string>("srcJetResPt")));
+      jetResPhiToken_ = esConsumes(edm::ESInputTag("", iConfig.getParameter<std::string>("srcJetResPhi")));
       rhoToken_ = consumes<double>(iConfig.getParameter<edm::InputTag>("srcRho"));
     }
 
@@ -106,9 +106,9 @@ namespace cms {
     edm::Handle<edm::View<reco::Jet>> inputJets;
     event.getByToken(jetToken_, inputJets);
 
-    JME::JetResolution resPtObj = JME::JetResolution::get(setup, jetResPtType_);
-    JME::JetResolution resPhiObj = JME::JetResolution::get(setup, jetResPhiType_);
-    JME::JetResolutionScaleFactor resSFObj = JME::JetResolutionScaleFactor::get(setup, jetSFType_);
+    JME::JetResolution resPtObj(setup.getData(jetResPtToken_));
+    JME::JetResolution resPhiObj(setup.getData(jetResPhiToken_));
+    JME::JetResolutionScaleFactor resSFObj(setup.getData(jetSFToken_));
 
     edm::Handle<double> rho;
     event.getByToken(rhoToken_, rho);

